File: docs/DSA/src/py/construct_binary_tree_from_preorder_and_inorder_traversal.py
from typing import Optional, List
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def _helper(self, left, right):
        if left > right:
            return

        val = self.preorder[self.pre_idx]
        self.pre_idx += 1
        mid = self.inorder[val]
        root = TreeNode(val)
        root.left = self._helper(left, mid-1)
        root.right = self._helper(mid+1, right)
        return root

    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.inorder = {v: idx for idx, v in enumerate(inorder)}
        self.pre_idx = 0
        self.preorder = preorder
        return self._helper(0, len(inorder)-1)

File: docs/DSA/src/go/construct_binary_tree_from_preorder_and_inorder_traversal.go
package main

func buildTree(preorder []int, inorder []int) *TreeNode {
	var helper func(left, right int) *TreeNode
	var preIdx int
	inorderMap := make(map[int]int)
	for idx, v := range inorder {
		inorderMap[v] = idx
	}
	helper = func(left, right int) *TreeNode {
		if left > right {
			return nil
		}

		val := preorder[preIdx]
		preIdx++
		mid := inorderMap[val]
		root := TreeNode{Val: val}
		root.Left = helper(left, mid-1)
		root.Right = helper(mid+1, right)
		return &root
	}

	return helper(0, len(inorder)-1)
}
